app/models/services/route_binding.rb

```diff
@@ -33,8 +35,36 @@ def required_parameters
       { route: route.uri }
     end
 
+    def last_operation
+      route_binding_operation
+    end
+
     def operation_in_progress?
-      false
+      !!route_binding_operation && route_binding_operation.state == 'in progress'
+    end
+
+    def terminal_state?
+      %w(succeeded failed).include? last_operation.state
+    end
+
+    def save_with_new_operation(attributes, last_operation)
+      RouteBinding.db.transaction do
+        self.lock!
+        set(attributes)
+        save_changes
+
+        if self.last_operation
+          self.last_operation.destroy
+        end
+
+        # it is important to create the service route binding operation with the service binding
+        # instead of doing self.service_route_binding_operation = x
+        # because mysql will deadlock when requests happen concurrently otherwise.
+        RouteBindingOperation.create(last_operation.merge(route_binding_id: self.id))
```

Comment:
I would probably rename the 2nd argument `new_operation` so that you don't have to deal with the overlap of `last_operation` and sometimes refer to `self.last_operation` when you mean the old one.

The only reason I spotted this was because it was drilled into me in my rubyist days that `self` is the default receiver, and so at first I assumed it was redundant here.